### ISSUE TITLE
In Wordpress Outline Assignment quickly becomes unwieldy / infeasible

### ISSUE THREAD

**Issue by AlwynBarry** [gantry/gantry5#1279]
In Wordpress all posts, pages, taxonomies, etc are considered as potential points for Outline assignment.  This rapidly results in an unwieldy Outline assignment page, and the site only needs a moderate number of posts before max_input_vars needs to be raised to a large number and the update script times out.

A solution would be to structure the Outline Assignment page.  Different 'categories' of post are already identified within separate boxes.  These could become separate assignment submission pages to reduce the number of updates at one time - the top-level 'categories' of post would be listed on the page, and selecting one could direct to a page or a modal dialog with the posts displayed on that page/dialog to select/deselect.  Within these, 'paging' of large numbers of any one post type would reduce the number of changes submitted at  once.

(see discussion here: http://www.rockettheme.com/forum/wordpress-theme-isotope/258649-major-fail-isotope-doesn-t-cope-with-outline-assignment-for-large-sites?start=0#1291819)

**Comment by newkind** [comment]
We're aware of this issue and we'll try to figure it out in a one or another way asap.

**Comment by w00fz** [comment]
Hi @AlwynBarry,
like Jakub has mentioned we are aware of the issue but we haven't had it tracked in here, so let me first thank you for that.

We have had a discussion internally about how to address it and decided we are going to resolve it in two steps. Let me recap it for you and for the records.

The first step is going to be resolving the `max_input_vars` problem, we reckon that is bad error and makes assignments potentially unusable when the amount of items is large. We are now going to only ever save 'checked' assignments rather than both checked and unchecked. This alone will cut down in the post request that happens during save and unlikely ever trigger the `max_input_vars` limit again.
This first step is going to be addressed ASAP and we are planning on releasing a new version of Gantry later this week.

The second step is going to be a bit more tricky to realize so we are going to postpone it for a later release. We do realize wordpress' assignable items can be a very huge amount, causing all kind of issues like making the Assignment page loading very slow or adding a lot of extra scrolling to the page which is not ideal.
This has actually been bugging us for long as well and we just could never find a better solution. What we agreed on doing now is to add pagination for each 'card' group. Each category/section will only list a fixed amount of items (say 10) and will display a pagination when needed.
All checked items will be displaying at the top of the list and will be always available for review and unchecking.
With this approach we think the page size will get trimmed quite a lot as well as making everything look cleaner and better organized.


I hope we could get to the step two soon enough, in the meantime stay tuned for the step one and more critical fix, I will notify in here when this happens and you will be able to try a [CI Build](gantry.org/downloads#ci-builds).

Thank you.

**Comment by AlwynBarry** [comment]
Thanks for looking at this ... it was a 'show-stopper' for me, and I'm amazed it hasn't been for others, though perhaps folk do most of their Outlines before a lot of content gets added, whereas I moved the content from Joomla beforehand.

The immediate fix of only submitting the changes is certainly a great step - thanks!!.

The longer term fix sounds good - making the initial outline assignment page a page where the 'cards' are identified and then you browse down from there to assign within the cards is the right solution.  The big problem that I foresee is if someone decides to change (say) all posts to a new outline ... with any blog this could quickly amount to hundreds and you'd quickly get to the max_input_vars problem arise again unless you captured the changes in an array of values that are passed in a single variable (for example).  If you have hundreds of posts it would be frustrating not to be able to make this change at the top-level 'switch' but have to go through the pages below the card level, so this is the obvious 'edge case' to cater for.  Apart from that, this solution should provide a much better approach to the problem.

For your info, although it is a particular problem with Wordpress, I did face a similar difficulty in Joomla where the problem was the script timing out rather than the max_input_variables when I changed from one outline to another for the majority of my pages.  I ended up having to change the original outline, which I didn't want to do because I wanted the opportunity to go back if the change wasn't effective for the users.  So, a solution will be useful for Joomla users too, I guess.

**Comment by w00fz** [comment]
Ok, part 1 of the issue has been now resolved and it is available for testing in our [CI Builds](https://github.com/gantry/gantry5/issues/gantry.org/downloads#ci-builds). We are still planning a new release of Gantry tomorrow or early next week.

Cheers!

**Comment by newkind** [comment]
About the page size - we had an implementation of the max-height of the cards and vertical scroll bar in each of them, so all cards would have the same size :

![isf4j51rlfcaol0xdayd9axgdhi5lb](https://cloud.githubusercontent.com/assets/1641534/15248207/3a544ea4-1919-11e6-8835-fcd58063445b.png)

Unfortunately we had to disable it because of a flexbox rendering bug in Chrome which has been there for the past 5 or 6 (sic!) versions! 

I just tested latetest Canary build and I can see that they finally started fixing it. It's still not ideal and causes some rendering issues, but hopefully it'll get sorted out when it goes stable.

![r5aeakaeitocc3ueysfobng9x3qjcf](https://cloud.githubusercontent.com/assets/1641534/15248197/2c8d34d4-1919-11e6-9a8d-d842111251d8.png)

**Comment by AlwynBarry** [comment]
Ok, I've tried the CI build and I can now assign layouts to pages!  The pause after each time an item is selected remains long, but at least I can work with layouts again!  Thanks so much!!  I can't test it extensively because I don't have that many different layouts, but at the moment it appears to work as expected.

**Comment by dudewithamood** [comment]
ETA on releasing the first part of the fix into the wild?

**Comment by w00fz** [comment]
Oh you must of missed it. It's already in the 5.2.18 release.

**Comment by mahagr** [comment]
Created a followup issue as 5.2.18 milestone has already been closed.

**Comment by dudewithamood** [comment]
Oh dear :-( I'm using 5.2.18 and get a 400 Bad Request. This is a site I'm migrating into Wordpress where there are four custom post types and well over 1000 posts, pages, menu items and categories combined. The max_input_vars is 1000.

**Comment by mahagr** [comment]
Where are you getting this 400 response? If assignments fail to save, it should have error which asks you to raise `max_input_vars` value. If the error is something else, you have likely ran into another issue, which may be unrelated to this issue.

**Comment by dudewithamood** [comment]
I get it when I try to save assignments and yes, it asks to raise max_input_vars.

**Comment by AlwynBarry** [comment]
It's working fine for me still ... updated to 5.2.18 and no errors.  It is still VERY slow to select, but seems to save Ok without error.  However, I guess it would still give an error if I was changing the default page, which would then change many allocations at a time.  I'm just careful to only change a few allocations each time.

My main problem now is that there are saved allocations that aren't being implemented when the theme is rendered.  I entered a new 'Contacts' layout and assigned the 'Contact Us' page and menu item to that layout.  All saved Ok, but the 'Contact Us' page is still using the Default layout.  With the current system of assignments I don't know if there is another assignment of some term or other that is overridding the settings, or whether it's an error in the allocation (when I go back to the allocation page, however, the allocations have been recorded correctly).  My thinking at the moment is that there is another setting that is overriding the page setting (maybe for the category term) because on some other pages I have had similar problems when there were a number of settings that could apply to the same page.  Sadly, Gantry provides no clear 'hierarchy' of settings that lets you see when you've changed all that are required.

Having used other Wordpress themes, in which you allocate to pages or posts in the Page editor, I have to say that the Gantry method is _much_ less intuitive for individual page/post allocation.  It works well with Joomla, but just doesn't really fit the Wordpress way of doing things.

**Comment by w00fz** [comment]
Even though Matias opened a separate issue with #1320, I'd rather reopen this issue and change milestone, since the discussion has been going on in here.

**Comment by newkind** [comment]
@AlwynBarry Menu Item Assignments are working by comparing current page URL to the menu item URL and they need to match in order to work. Are you sure that you don't have any extra parameters in the menu item url or maybe you're using any language switcher plugin ? I'm sure there must be cause why it wouldn't work for you.

**Comment by AlwynBarry** [comment]
I've checked and the page URL and menu URL are the same ... no extra parameters, and no language switcher involved.  The only thing that is different is that permalinks are set to %category%/%postname% but the menu picks that up fine - join-us/contact-us.  The page slug is contact-us, but I've assigned both the page and the menu item to the outline.  (I've also tried all combinations of setting either or both of these!).

Looking in the assignments.yaml file for the assigned Outline we have:
context: {  }
menu:
  main-menu:
    3027: true
post:
  page:
    3024: true
taxonomy: {  }
archive: {  }

Neither of these numbers appear in the default assignments.yaml and so the assignments have been made correctly by the script.

Querying the site with URL .../?p=3024 gives the page  join-us/contact-us/  and looking in the database 3024 is the post with postname contact-us and database item 3027 is the nav_menu_item post type.

So, I can't see any problem with the assignment script or the name or the menu entry.  Any other ideas?  I guess you can see why I said that the way Gantry does the assignments is rather unintuitive in Wordpress because there are potentially category assignments that could (I guess) override the page assignments ... but there's no way of telling if that's whats happening.

So, in the redesign, the 'hierarchy' of assignment decisions needs to also be identified, I think.

But all the above is slightly 'off topic'.  The original problem was the problem with max_input_vars and the unweildy nature of the assignments page.  This is now a problem with the transparency of how the assignments operate through to what is seen in the page rendering, which you may see as 'off topic'.

**Comment by mahagr** [comment]
I remembered that there's issue for this, but maybe its closed as I cannot find it.

You can see the decision by following code:
```
$assignments = new Gantry\Framework\Assignments;
print_r($assignments->scores());
```

**Comment by newkind** [comment]
Please check what @mahagr wrote and if this won't give you a hint, you can ping me in Gitter and I can try to find a solution for you.

**Comment by nikola3244** [comment]
@newkind, scrolling problem disappears when you remove `-webkit-backface-visibility: hidden;` and `backface-visibility: hidden;` from `<label>`.
I don't see any difference when it's there and when it's not.

In Chromium 51.0.2704.84 (64-bit)
https://drive.google.com/file/d/0B8fB_QtQf14TQUFEN1FaaGM5eW8/view?usp=sharing

**Comment by w00fz** [comment]
Probably best to test with CI Builds, there's been a lot of changes in the SCSS about assignments, I might already have fixed that.

**Comment by newkind** [comment]
@nikola3244 this should be fixed already in latest CI and Chrome stable - 51.0.2704.103 (64-bit)

**Comment by nikola3244** [comment]
True, it's fixed.

**Comment by AlwynBarry** [comment]
Finally got chance to test the suggestion of @mahagr 

The output for the outline choice was:

<!--Array
(
    [sbc_contact] => 3
)
Array
(
    [sbc_contact] => Array
        (
            [menu] => Array
                (
                    [main-menu] => Array
                        (
                            [3027] => 3
                        )

                )

        )

)

But, it's still displaying the default template!

But, now a bigger problem ... I've moved to the production site and I don't have access to the max_vars setting and now, even with the up to date Gantry 5, it always gives me a max_vars error.

Sticking with Gantry to get the site going, but can't stick with Gantry 5 long term.  Such a pity - it worked like a dream in Joomla!

**Comment by mahagr** [comment]
You're right, the fix doesn't seem to work in WP, I just checked it out. It works in Joomla, but not in WP. 

CC: @w00fz

**Comment by w00fz** [comment]
Sorry, you are right, for some reason my implementation was only applied to Joomla. 
I fixed this now in the currente develop, you can try the latest CI Build and it should be working as discussed above.

**Comment by mahagr** [comment]
I don't think that the issue is fully fixed (nor that it was intended to fix from the commit) -- we still need to start using ajax to make the page easier for the poor browsers.

**Comment by AlwynBarry** [comment]
I've applied the CI and can change the assignments without error.  But I get a 'White screen of death' on the site with the CI build, though Ok again when I restore the current stable.  No source HTML available to check, and I don't have time just now to investigate what is causing the WSOD.

**Comment by mahagr** [comment]
For me everything works, so it may be hard to track down. Which version of PHP are you using? Do you have opcache enabled? Are you seeing anything in Apache error.log?

**Comment by nikola3244** [comment]
@AlwynBarry Also, enable [WP debug mode](https://codex.wordpress.org/Debugging_in_WordPress). You might see some errors.

**Comment by mahagr** [comment]
iThis Issue affects Joomla too: #1768

**Comment by AmibeWebsites** [comment]
A thought on this, as a temporary workaround... it is very seldom (probably never) that I assign a layout to an individual post - a page yes (talking WordPress here), but not a post. Posts are like news items or products and so I'm far more likely to assign a layout to a category or post type.

And since it's mostly posts (don't often have over 1000 categories) that cause the sloooooooooowness, could there be a setting "Display posts for assignment" set to False by default?

What are others experiences with assigning to individual post?

**Comment by AmibeWebsites** [comment]
Dirty, but fast workaround... copy the file admin/templates/pages/configurations/assignments/assignments.html.twig to your theme add the following condition within appropriate _for_ loop:
`{% if not name or name not in ['post', 'shop_order', 'shop_coupon', 'shop_webhook', 'link_category', 'attachment', 'product'] %}`

**Comment by mahagr** [comment]
I just added Gantry options for removing posts and pages (= all but posts) into CI build, please test it out.

**Comment by igor-gp** [comment]
Wow i have just hit this issue myself with a website and it is sooo slow. It is not usable. 

I am in WordPress, latest version of everything. @newkind any way we can speed this up so we can use it? 

I have turned off Posts in the settings but that hasn't really helped.

**Comment by newkind** [comment]
Nothing really we can do about it without some bigger changes. Not sure when that happens.

**Comment by alexbangle** [comment]
Has anyone come up with a work around for this?  Maybe some code that could be used to only show certainly areas in assignments?  The issue I'm having is with an update to Yoast SEO it now fills the database with private taxonomies that it uses for it's new insights features.  This has created 1,000's and soon 10,000's of custom assignments in the gantry menu making it unresponsive.  

I tried a few of the items above but they ended up just causing an error on the assignments page.  I've reached out to the plugin author as well and they we tried a few thing with the plugin but aside from uninstalling it (which we can't) it seems we've locked up the assignments page until it's fixed.

Let me know if anyone has some thoughts on this, I'm willing to try things on our development site to try and find a work around.
